File: algorithms_in_graphs/week4/_efcbd6ce753b5e2256a932c991fdca28_10_paths_in_graphs_starter_files_2/shortest_paths/shortest_paths.cpp
#include <iostream>
#include <limits>
#include <vector>
#include <queue>

using std::vector;
using std::queue;
using std::pair;
using std::priority_queue;
using std::numeric_limits;

void shortest_paths(vector<vector<int> > &adj, vector<vector<int> > &cost, int s, vector<long long> &distance, vector<int> &reachable, vector<int> &shortest) {
  //write your code here
  vector<int> negative_set;
  	reachable[s] = 1;
  	distance[s] = 0;
    int current_index,next_index;
  	for (int i = 0; i < adj.size(); i++)
  	{
  		for (int j = 0; j < adj.size(); j++)
  		{
  			current_index = j;
  			if (distance[current_index]
  					!= numeric_limits<long long>::max())
  			{
  				for (int k = 0; k < adj[j].size(); k++)
  				{
  					next_index = adj[j][k];
  					if (distance[next_index]
  							> distance[current_index] + cost[current_index][k])
  					{
  						if (i == adj.size() - 1)
  						{
  							negative_set.push_back(next_index);
  						}
  						distance[next_index] = distance[current_index]
  								+ cost[current_index][k];
  						reachable[next_index] = 1;
  					}
  				}
  			}

  		}
  	}
    int node_index,current_node,adjNode;
  	for (int i = 0; i < negative_set.size(); i++)
  	{
  		node_index = negative_set[i];
  		if (shortest[node_index])
  		{
  			queue<int> visiting_queue;
  			visiting_queue.push(node_index);
  			while (!visiting_queue.empty())
  			{
  				current_node = visiting_queue.front();
  				shortest[current_node] = 0;
  				visiting_queue.pop();
  				for (int i = 0; i < adj[current_node].size(); i++)
  				{
  					adjNode = adj[current_node][i];
  					if (shortest[adjNode])
  					{
  						visiting_queue.push(adjNode);
  						shortest[adjNode] = 0;
  					}
  				}
  			}
  		}

  	}

}

int main() {
  int n, m, s;
  std::cin >> n >> m;
  vector<vector<int> > adj(n, vector<int>());
  vector<vector<int> > cost(n, vector<int>());
  for (int i = 0; i < m; i++) {
    int x, y, w;
    std::cin >> x >> y >> w;
    adj[x - 1].push_back(y - 1);
    cost[x - 1].push_back(w);
  }
  std::cin >> s;
  s--;
  vector<long long> distance(n, std::numeric_limits<long long>::max());
  vector<int> reachable(n, 0);
  vector<int> shortest(n, 1);
  shortest_paths(adj, cost, s, distance, reachable, shortest);
  for (int i = 0; i < n; i++) {
    if (!reachable[i]) {
      std::cout << "*\n";
    } else if (!shortest[i]) {
      std::cout << "-\n";
    } else {
      std::cout << distance[i] << "\n";
    }
  }
}

File: algorithms_in_graphs/week4/_efcbd6ce753b5e2256a932c991fdca28_10_paths_in_graphs_starter_files_2/shortest_paths/shortest_paths.py
#Uses python3

#import sys
import Queue


def shortest_paths(adj, cost, s, distance, reachable, shortest):
    #write your code here
    #pass
    distance[s] = 0
    reachable[s] = 1
    q = Queue.Queue()
    for i in range(len(adj)):
        for u in range(len(adj)):
            for v in adj[u]:
                v_index = adj[u].index(v)
                if distance[u] != 10**19 and distance[v] > distance[u] + cost[u][v_index]:
                    distance[v] = distance[u] + cost[u][v_index]
                    reachable[v] = 1
                    if i == len(adj) - 1:
                        q.put(v)

    visited = [0] * len(adj)
    while not q.empty():
        u = q.get()
        visited[u] = 1
        if u != s:
            shortest[u] = 0
        for v in adj[u]:
            if visited[v] == 0:
                q.put(v)
                visited[v] = 1
                shortest[v] = 0
    distance[s] = 0



if __name__ == '__main__':
    n, m = map(int, raw_input().split())
    adj = [[] for _ in range(n)]
    cost = [[] for _ in range(n)]
    for i in range(m):
        a, b, w = map(int, raw_input().split())
        adj[a - 1].append(b - 1)
        cost[a - 1].append(w)
    s = int(raw_input())

    s -= 1
    distance = [10**19] * n
    reachable = [0] * n
    shortest = [1] * n
    shortest_paths(adj, cost, s, distance, reachable, shortest)
    for x in range(n):
        if reachable[x] == 0:
            print('*')
        elif shortest[x] == 0:
            print('-')
        else:
            print(distance[x])
